src/librustdoc/html/render/mod.rs

```diff
@@ -737,42 +737,85 @@ fn render_impls(
-    use crate::formats::item_type::ItemType::*;
+/// Build a (possibly empty) `href` attribute (a key-value pair) for the given associated item.
+fn assoc_href_attr(it: &clean::Item, link: AssocItemLink<'_>, cx: &Context<'_>) -> String {
+    let name = it.name.unwrap();
+    let item_type = it.type_();
 
-    let name = it.name.as_ref().unwrap();
-    let ty = match it.type_() {
-        Typedef | AssocType => AssocType,
-        s => s,
-    };
+    let href = match link {
+        AssocItemLink::Anchor(Some(ref id)) => Some(format!("#{}", id)),
+        AssocItemLink::Anchor(None) => Some(format!("#{}.{}", item_type, name)),
+        AssocItemLink::GotoSource(did, provided_methods) => {
+            // We're creating a link from the implementation of an associated item to its
+            // declaration in the trait declaration.
+            let item_type = match item_type {
+                // For historical but not technical reasons, the item type of methods in
+                // trait declarations depends on whether the method is required (`TyMethod`) or
+                // provided (`Method`).
+                ItemType::Method | ItemType::TyMethod => {
+                    if provided_methods.contains(&name) {
+                        ItemType::Method
+                    } else {
+                        ItemType::TyMethod
+                    }
+                }
+                // For associated types and constants, no such distinction exist.
+                item_type => item_type,
+            };
 
-    let anchor = format!("#{}.{}", ty, name);
-    match link {
-        AssocItemLink::Anchor(Some(ref id)) => format!("#{}", id),
-        AssocItemLink::Anchor(None) => anchor,
-        AssocItemLink::GotoSource(did, _) => {
-            href(did.expect_def_id(), cx).map(|p| format!("{}{}", p.0, anchor)).unwrap_or(anchor)
+            match href(did.expect_def_id(), cx) {
+                Ok((url, ..)) => Some(format!("{}#{}.{}", url, item_type, name)),
+                // The link is broken since it points to an external crate that wasn't documented.
+                // Do not create any link in such case. This is better than falling back to a
+                // dummy anchor like `#{item_type}.{name}` (which used to happen in older versions)
+                // that in most cases indeed coincides with the ID of the impl item. In general
+                // however, this dummy would be incorrect: If the type with the trait impl also
+                // has an inherent impl with an assoc. item of the *same* name as one of the
+                // assoc. items of the trait, the dummy would link to that one even though those
+                // items are distinct! In this scenario, the actual ID of the impl item would be
+                // `#{item_type}.{name}-{n}` for some number `n`, a disambiguator.
+                Err(HrefError::DocumentationNotBuilt) => None,
```

Comment:
> This is better than falling back to a dummy anchor like `#{item_type}.{name}` (which used to happen in older versions)

It used to happen with associated types and constants (through the use of the now removed `naive_assoc_href` which was replaced by this function – `assoc_href_attr`) and it used to happen for `Method`s (not `TyMethod`s) however. Since this line used to look like (not literally but in spirit):

```diff
- Err(HrefError::DocumentationNotBuilt) => None,
+ Err(HrefError::DocumentationNotBuilt) if item_type == ItemType::TyMethod => None,
```

See also the test `extern-default-method` I had to modify.